cmd/gator/validate/validate.go

```diff
@@ -0,0 +1,200 @@
+  cat manifest.yaml | gator validate
+
+  # Output structured violations data
+  gator validate --filename="manifest.yaml" --json`
+)
+
+var Cmd = &cobra.Command{
+	Use:     "validate",
+	Short:   "validate resources against templates and constraints",
+	Example: examples,
+	Run:     run,
+}
+
+var (
+	flagFilenames []string
+	flagYAML      bool
+	flagJSON      bool
+)
+
+const (
+	flagNameFilename = "filename"
+	flagNameJSON     = "json"
+	flagNameYAML     = "yaml"
+)
+
+func init() {
+	Cmd.Flags().StringArrayVarP(&flagFilenames, flagNameFilename, "f", []string{}, "a file or directory containing kubernetes resources.  Can be specified multiple times.  Cannot be used in tandem with stdin.")
+	Cmd.Flags().BoolVar(&flagYAML, flagNameYAML, false, "print validation information as yaml.")
+	Cmd.Flags().BoolVar(&flagJSON, flagNameJSON, false, "print validation information as json.")
+
+	errLog = log.New(os.Stderr, "", 0)
+	outLog = log.New(os.Stdout, "", 0)
+}
+
+func run(cmd *cobra.Command, args []string) {
+	var unstrucs []*unstructured.Unstructured
+
+	// check if stdin has data
+	stdinfo, err := os.Stdin.Stat()
+	if err != nil {
+		errLog.Fatalf("getting info for stdout: %s", err)
+	}
+
+	// using stdin in combination with flags is not supported
+	if stdinfo.Size() > 0 && len(flagFilenames) > 0 {
+		errLog.Fatalf("stdin cannot be used in combination with %q flag", flagNameFilename)
+	}
+
+	// if no files specified, read from Stdin
+	if stdinfo.Size() > 0 {
+		us, err := readYAMLSource(os.Stdin)
+		if err != nil {
+			errLog.Fatalf("reading from stdin: %s", err)
+		}
+		unstrucs = append(unstrucs, us...)
+	} else if len(flagFilenames) > 0 {
+		// normalize directories by listing their files
+		normalized, err := normalize(flagFilenames)
+		if err != nil {
+			errLog.Fatalf("normalizing: %s", err)
+		}
+
+		for _, filename := range normalized {
+			file, err := os.Open(filename)
+			if err != nil {
+				errLog.Fatalf("opening file %q: %s", filename, err)
+			}
+
+			us, err := readYAMLSource(bufio.NewReader(file))
+			if err != nil {
+				errLog.Fatalf("reading from file %q: %s", filename, err)
+			}
+			file.Close()
+
+			unstrucs = append(unstrucs, us...)
+		}
+	} else {
+		errLog.Fatalf("no input data: must include data via either stdin or the %q flag", flagNameFilename)
+	}
+
+	responses, err := validate.Validate(unstrucs)
+	if err != nil {
+		errLog.Fatalf("auditing objects: %v\n", err)
+	}
+
+	results := responses.Results()
+
+	if flagYAML && flagJSON {
+		errLog.Fatalf("cannot use --%s and --%s in combination", flagNameJSON, flagNameYAML)
+	}
+
+	if flagJSON {
+		b, err := json.MarshalIndent(results, "", "    ")
+		if err != nil {
+			errLog.Fatalf("marshaling validation json results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	}
+
+	if flagYAML {
+		b, err := yaml.Marshal(results)
+		if err != nil {
+			errLog.Fatalf("marshaling validation yaml results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	}
+
+	if len(results) > 0 {
```

Comment:
How are we handling different enforcement actions, like `audit`, or `warn`?

Reply:
> How are we handling different enforcement actions, like dryrun, or warn? 

An interesting question.  The utility of `dryrun` and `warn` are obvious on a running cluster where we don't want to get in the way of daily operations as we develop/test policy.  Shift-left could be different, as we don't have to worry about breaking our entire org's deploy jobs with a line of bad rego.

Leveraging `dryrun` and `warn` in `gator validate` would likely mean omitting messages related to constraints with these actions enabled.  This _would_ put `gator validate` more inline with the webhook.  i.e, an object that gets past the webhook would get past `gator validate`.  This makes `gator validate` _less_ like the Audit process.  Results that would show up in Audit under a constraint's `status` section would now be omitted from results.

I can see an argument for both, given that some users may be more "audit centric" vs. "webhook centric".  

But, I think we can help both users by *omitting* `dryrun` and `warn` violations from `gator validate` output.  I believe this will work well because users who only use g8r's Audit functionality have no need for `enforcementAction`.  Their experience of `gator validate` will match their Audit experience since they're unlikely to engage with this feature.

Users of `dryrun` and `warn` will likely be webhook users, as those modes' purpose is to _exempt the webhook from the constraint_.  If users have the understanding to enable the feature, they'll likely want `gator validate` to behave the same way.

That said, an organization may choose to enable a policy in shift-left _before_ it goes into `deny` on the cluster as a way to gradually ramp up teams on the new policy.  We could meet such a use-case with another flag like `--force-deny`.  Given that's a bit of a niche scenario, I think we could leave it out of MVP and see if users request it.

> What about unknown enforcement actions?

It [looks like](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/util/enforcement_action.go#L33) g8r won't accept a Constraint that includes an enforcementAction it doesn't support.  Perhaps we should do the same?  I could call out to this same package to make sure `gator validate` stays up to date with g8r if we add another supported action.

Reply:
enforcementActions are a bit weird in that they reflect the capabilities of the enforcement point. For example, K8s validating webhooks didn't originally support warnings. If other enforcement points recognize actions that G8r doesn't and the validation of `enforcementAction` gets in the way, it can be disabled via flag. TL;DR, we shouldn't assume we're only looking at constraints intended for enforcement by G8r.

TBH I think `enforcementActions` are missing the ability to specify different enforcement actions based on the current enforcement point, but that's a future problem.

For now, we should consider the value of `gator validate` relative to validating webhooks and audit and use that to build its value as a standalone product, vs. designing by analogy.

We have two levers to work with here:

- return code (0 for allow, then we can define other codes as necessary)
- output (stdout vs. stderr, output format, omitting values)

I'd say: return all violations, but only return non-zero if there is a deny-level violation

If we wanted to make it customizable (say users want to reject on warnings so that no future deprecated changes are applied), we could add a flag that classifies the set of enforcement actions that lead to rejection. This would also allow users to specify the significance of unknown enforcement actions.

Might be fun to play around with color-coding, but that's beyond MVP.

Reply:
After chats with Max, I've implemented the following:

- If any of the violated constraints have `enforcementAction: deny` (the default value), the exit code returned is `1`
- Any other `enforcementAction` (`dryrun`, `warn`, `foobar`, etc.) will cause a `0` exit code to be returned
- violation results are printed to stdout, regardless of `enforcementAction`